ntp-client/src/Network/NTP/Client.hs

```diff
@@ -65,6 +56,54 @@ data NtpClientSettings = NtpClientSettings
     }
 
 data NtpClient = NtpClient
+    { ntpGetStatus        :: IO NtpStatus -- Use STM here ?
+      -- ^ Query the current status (non-blocking).
+    , ntpActNow           :: IO ()
+      -- ^ Bypass all internal threadDelays and trigger a new ntp query.
+    , ntpForceCheck       :: IO NtpStatus
+      -- ^ Perform a ntp query and wait for the results (blocking with timeout).
+    , ntpAbort            :: IO () --todo/discuss
```

Comment:
do you mean a clean shutdown of the service? As a first step I would test if one can shutdown it with an async exception, e.g. do all the threads are shut down properly?

Reply:
Good point I'll check out the async exception.

Reply:
This does not work:
```
test :: IO ()
test = do
    t <- async $ startNtpClient (contramapM (return . show) stdoutTracer) settings
    threadDelay 30000000
    putStrLn "trying to kill the client"
    cancel t
```
Also an explicit function may be better.
An other option may be to use a proper with-style wrapper
`withNtpClient :: Tracer IO NtpTrace -> NtpClientSettings -> (NtpClient -> IO a) -> IO a`

Reply:
yes, but this will for sure succeed.  The question is if all the threads that `startNtpClient` starts will be terminated as well. In Haskell if you start a thread it can out live the parent thread. One needs to analyse the code.

Reply:
btw, `startNtpClient` is already using `async` to fork the client, so there's no need to to fork it again as you do you the snippet above.